code/game/objects/items/mail.dm

```diff
@@ -0,0 +1,262 @@
+	var/list/generic_goodies = list(
+		/obj/item/stack/spacecash/c100 = 30,
+		/obj/item/stack/spacecash/c200 = 20,
+		/obj/item/stack/spacecash/c500 = 10,
+		/obj/item/stack/spacecash/c1000 = 5,
+		/obj/item/stack/spacecash/c10000 = 1
+	)
+	// Overlays (pure fluff)
+	var/postmarked = TRUE
+	var/stamped = TRUE
+	var/list/stamps = list()
+	var/stamp_max = 1
+	var/stamp_offset_x = 0
+	var/stamp_offset_y = 2
+
+	///mail will have the color of the department the recipient is in.
+	var/static/list/department_colors
+
+/obj/item/mail/envelope
+	name = "envelope"
+	icon_state = "mail_large"
+	goodie_count = 2
+	stamp_max = 2
+	stamp_offset_y = 5
+
+/obj/item/mail/Initialize()
+	. = ..()
+
+	department_colors = list(
+	ACCOUNT_CIV = COLOR_WHITE,
+	ACCOUNT_ENG = COLOR_PALE_ORANGE,
+	ACCOUNT_SCI = COLOR_PALE_PURPLE_GRAY,
+	ACCOUNT_MED = COLOR_PALE_BLUE_GRAY,
+	ACCOUNT_SRV = COLOR_PALE_GREEN_GRAY,
+	ACCOUNT_CAR = COLOR_BEIGE,
+	ACCOUNT_SEC = COLOR_PALE_RED_GRAY,
+	)
+
+	// Icons
+	// Add some random stamps.
+	if(stamped == TRUE)
+		var/N = rand(1, stamp_max)
+		for(var/i = 1, i <= N, i++)
+			var/X = rand(2, 6)
+			stamps += list("stamp_[X]")
+	update_icon()
+
+/obj/item/mail/update_icon()
+	cut_overlays()
+
+	var/bonus_stamp_offset = 0
+	for(var/S in stamps)
+		var/image/SI = image(
+			icon = icon,
+			icon_state = S,
+			pixel_x = stamp_offset_x,
+			pixel_y = stamp_offset_y + bonus_stamp_offset
+		)
+		// Stops postmarks from inheriting letter color.
+		// http://www.byond.com/docs/ref/#/atom/var/appearance_flags
+		SI.appearance_flags |= RESET_COLOR
+		add_overlay(SI)
+		bonus_stamp_offset -= 5
+
+	if(postmarked == TRUE)
+		var/image/PMI = image(
+			icon = icon,
+			icon_state = "postmark",
+			pixel_x = stamp_offset_x + rand(-3, 1),
+			pixel_y = stamp_offset_y + rand(bonus_stamp_offset + 3, 1)
+		)
+		PMI.appearance_flags |= RESET_COLOR
+		add_overlay(PMI)
+
+/obj/item/mail/attackby(obj/item/W, mob/user, params)
+	// Destination tagging
+	if(istype(W, /obj/item/dest_tagger))
+		var/obj/item/dest_tagger/O = W
+
+		if(sortTag != O.currTag)
+			var/tag = uppertext(GLOB.TAGGERLOCATIONS[O.currTag])
+			to_chat(user, "<span class='notice'>*[tag]*</span>")
+			sortTag = O.currTag
+			playsound(loc, 'sound/machines/twobeep_high.ogg', 100, TRUE)
+
+/obj/item/mail/attack_self(mob/user)
+	if(recipient && user != recipient)
+		to_chat(user, "<span class='notice'>You can't open somebody else's mail! That's <em>illegal</em>!</span>")
+		return
+
+	to_chat(user, "<span class='notice'>You start to unwrap the package...</span>")
+	if(!do_after(user, 15, target = user))
+		return
+	user.temporarilyRemoveItemFromInventory(src, TRUE)
+	unwrap_contents()
+	for(var/X in contents)
+		var/atom/movable/AM = X
+		user.put_in_hands(AM)
+	playsound(src.loc, 'sound/items/poster_ripped.ogg', 50, TRUE)
+	qdel(src)
+
+/obj/item/mail/proc/unwrap_contents()
+	for(var/obj/I in src.GetAllContents())
+		SEND_SIGNAL(I, COMSIG_STRUCTURE_UNWRAPPED)
```

Comment:
```suggestion
	for(var/i in GetAllContents())
		SEND_SIGNAL(i, COMSIG_STRUCTURE_UNWRAPPED)
```

Reply:
This did not compile without the `obj/`.

```
code\game\objects\items\mail.dm:123:error: i.comp_lookup: undefined var
code\game\objects\items\mail.dm:123:error: i.comp_lookup: undefined var
code\game\objects\items\mail.dm:123:error: i._SendSignal: undefined proc
```

Reply:
```suggestion
	for(var/I in GetAllContents())
		var/atom/unwrapped_atom = i
		SEND_SIGNAL(unwrapped_atom, COMSIG_STRUCTURE_UNWRAPPED)
```

the reason this is better is because src. is implied, and because if you do a typed for loop; byond will run an istype() for each object in the loop. While this can be beneficial in some cases where you would do an is type anyways. getallcontents only returns atoms, and thus istyping again is not really useful.

Reply:
That makes sense, thank you